File: implementations/sson.java
import java.util.HashMap;

public class sson
{
  public class parsingErrorException extends Exception { public ParsingErrorException(String errMsg) { super(errMsg); }}

  public HashMap<String, HashMap<String, String>> trySetObjects(String[] rawObjectData) throws ParsingErrorException
  {
    boolean readingObject = false, readingDefault = false;

    int lineCount = 0;
    var currentObject = "";

    var defaultValues = new HashMap<String, HashMap<String, String>>();
    var objects = new HashMap<String, HashMap<String, String>>();

    for (String str : rawObjectData) {

      ++lineCount;

      if (str.startsWith("#") || str.length() == 0) continue;

      if (str.startsWith(".")) {

        if (!readingObject)
          throw new parsingErrorException(String.format("%s at line %d is supposed to be a property, however it is cut off from its parent object. You probably misplaced a ; just before that line.\n", str, lineCount));

        
        var keyValuePair = str.split("=", 2);

        if (keyValuePair.length < 2)
          throw new parsingErrorException(String.format("Expected a value after %s at line %d; property cannot be empty.\n", str, lineCount));


        keyValuePair[0] = keyValuePair[0].substring(1);

        if (keyValuePair[1].endsWith(";")) 
          keyValuePair[1] = keyValuePair[1].substring(0, keyValuePair[1].length() - 1);

        keyValuePair[0] = keyValuePair[0].trim();
        keyValuePair[1] = keyValuePair[1].trim();

        if (readingDefault) {

          if (!defaultValues.containsKey(currentObject))
            defaultValues.put(currentObject, new HashMap<String, String>());

          if (!defaultValues.get(currentObject).containsKey(keyValuePair[0]))
            defaultValues.get(currentObject).put(keyValuePair[0], keyValuePair[1]);

          else defaultValues.get(currentObject).replace(keyValuePair[0], keyValuePair[1]);
        }

        else {

          if (!objects.containsKey(currentObject))
            objects.put(currentObject, new HashMap<String, String>());

          if (!objects.get(currentObject).containsKey(keyValuePair[0]))
            objects.get(currentObject).put(keyValuePair[0], keyValuePair[1]);

          else objects.get(currentObject).replace(keyValuePair[0], keyValuePair[1]);
        }
      }

      else {

        if (readingObject)
          throw new ParsingErrorException(String.format("Expected a ; before line " + lineCount));

        if (str.startsWith("default")) {

          readingObject = true;
          readingDefault = true;

          // removes the default part of the string.
          currentObject = str.substring(7).trim();
        }

        else {

          var cleanStr = str.trim();

          currentObject = String.format("%s_%s", cleanStr, lineCount);
          readingObject = true;

          if (defaultValues.containsKey(cleanStr)) {

            objects.put(currentObject, new HashMap<String, String>());

            for (var property : defaultValues.get(cleanStr).entrySet())
              objects.get(currentObject).put(property.getKey(), property.getValue());
          }
        }
      }

      if (str.endsWith(";")) {

        if (!readingObject)
          System.err.println("Redundant ; at line " + lineCount);

        readingObject = false;
        readingDefault = false;
      }
    }

    return objects;
  }
}

File: implementations/sson.php
<?php 

    function try_get_objects($raw_object_data)
    {
        if (!is_array($raw_object_data))
            return 'This function expects the parameter to be an array';

        $reading_object = false;
        $reading_default = false;

        $line_count = 0;
        $current_object = "";

        $objects = [];
        $default_values = [];

        foreach ($raw_object_data as $str)
        {
            ++$line_count;

            if (str_starts_with($str, '#') || $str === '') continue;

            if (str_starts_with($str, '.'))
            {
                if (!$reading_object)
                    return "$str at line $line_count is supposed to be a property, however it is cut off from its parent object. You probably misplaced a ; just before that line.";

                $key_value_pair = explode('=', $str, 2);

                if (count($key_value_pair) < 2)
                    return "expected a value after $str at line $line_count; property cannot be empty.";

                // remove the dot at the beginning of the attribute
                $key_value_pair[0] = substr($key_value_pair[0], 1);

                if (str_ends_with($key_value_pair[1], ';'))
                    $key_value_pair[1] = substr($key_value_pair[1], 0, strlen($key_value_pair[1]) - 1);

                $key_value_pair[0] = trim($key_value_pair[0]);
                $key_value_pair[1] = trim($key_value_pair[1]);

                if ($reading_default)
                {
                    if (!array_key_exists($current_object, $default_values))
                        $default_values[$current_object] = [];

                    $default_values[$current_object][$key_value_pair[0]] = $key_value_pair[1];
                }

                else
                {
                    if (!array_key_exists($current_object, $objects))
                        $objects[$current_object] = [];

                    $objects[$current_object][$key_value_pair[0]] = $key_value_pair[1];
                }
            }

            else
            {
                if ($reading_object) return "expected a ; before line $line_count";

                if (str_starts_with($str, 'default'))
                {
                    $reading_object = true;
                    $reading_default = true;

                    // removes the default part of the string
                    $current_object = trim(substr($str, 7));
                }

                else
                {
                    $clean_str = trim($str);
                    $reading_object = true;

                    if (array_key_exists($clean_str, $default_values))
                    {
                        $objects[$current_object] = [];

                        foreach ($default_values[$clean_str] as $property => $value)
                            $objects[$current_object][$property] = $value;
                    }
                }
            }

            if (str_ends_with($str, ';'))
            {
                $reading_object = false;
                $reading_default = false;
            }
        }

        return $objects;
    }
?>
